{
 "cells": [
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "<span style=\"font-size: 4em;\">Welcome to PyTreeMotif</span>\n",
    "<p></p>\n",
    "\n",
    "#### an enumerative motif finding algorithm that makes trees"
   ],
   "id": "18e78d85d57d5ad0"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "# IMPORTS \n",
    "\n",
    "Dependencies, can't live with em, can't live without em"
   ],
   "id": "2376f43055f51e0f"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-04-21T02:05:57.848416Z",
     "start_time": "2024-04-21T02:05:57.833834Z"
    }
   },
   "cell_type": "code",
   "source": [
    "import os\n",
    "import sys\n",
    "\n",
    "from collections import deque\n",
    "import random\n",
    "from tqdm import tqdm\n",
    "import re"
   ],
   "id": "95847246d7c1e8cb",
   "outputs": [],
   "execution_count": 1
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    " # CLASSES \n",
    " \n",
    "### Define custom classes needed for PyTreeMotif"
   ],
   "id": "28b20b1751619764"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "### Sequence_Node\n",
    "\n",
    "* nodes representing lmer sequences similar to a reference pair <p></p>\n",
    "\n",
    "* these are candidates from which tree nodes will be selected <p></p>\n",
    " \n",
    "* records sequence, position, and the input sequence in which it was found <p></p>"
   ],
   "id": "47fefe7820b4efad"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-04-21T02:05:58.678982Z",
     "start_time": "2024-04-21T02:05:58.675766Z"
    }
   },
   "cell_type": "code",
   "source": [
    "class Sequence_Node:\n",
    "    def __init__(self, seq, id, pos):\n",
    "        self.seq = seq\n",
    "        self.id = id\n",
    "        self.pos = pos"
   ],
   "id": "a3e874f49cc80b3a",
   "outputs": [],
   "execution_count": 2
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "### Sequence_Node_Set\n",
    "\n",
    "* a set of Sequence_Nodes from one input sequence that are similar to a reference pair\n"
   ],
   "id": "44e36c036d798c4d"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-04-21T02:05:59.604607Z",
     "start_time": "2024-04-21T02:05:59.600834Z"
    }
   },
   "cell_type": "code",
   "source": [
    "class Sequence_Node_Set:\n",
    "    def __init__(self, id):\n",
    "        self.id = id\n",
    "        self.members = set()\n",
    "        \n",
    "    def add_member(self, member):\n",
    "        self.members.add(member)"
   ],
   "id": "c657858a1e656fd4",
   "outputs": [],
   "execution_count": 3
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "### Reference_Pair\n",
    "\n",
    "* reference pair of lmer sequences differing by at most 2d <p></p>\n",
    "\n",
    "* one is from input sequence 0, the other from input sequence 1 <p></p>\n",
    "\n",
    "* reference pairs are the basis on which all other possible nodes are selected <p></p>\n",
    "\n",
    "* contains a pair of sequences, their positions, and an id number for the pair <p></p>\n",
    "\n",
    "* contains a list of Sequence_Node_Sets, one for each of the remaining input sequences <p></p>\n",
    "\n",
    "* after tree construction, contains a list of trees built from this reference pair <p></p>"
   ],
   "id": "ef34019e8dfe50ea"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-04-21T02:06:00.349931Z",
     "start_time": "2024-04-21T02:06:00.346201Z"
    }
   },
   "cell_type": "code",
   "source": [
    "class Reference_Pair:\n",
    "    def __init__(self, pair_id, seq1, seq2, pos1, pos2, num_seq):\n",
    "        self.pair_id = pair_id\n",
    "        self.seq1 = seq1\n",
    "        self.seq2 = seq2\n",
    "        self.pos1 = pos1\n",
    "        self.pos2 = pos2\n",
    "        \n",
    "        self.node_sets = [] \n",
    "        for i in range(2, num_seq):\n",
    "            self.node_sets.append(Sequence_Node_Set(i))\n",
    "            \n",
    "        self.tree_list = []\n",
    "\n",
    "        \n",
    "    "
   ],
   "id": "291f106cddb4d531",
   "outputs": [],
   "execution_count": 4
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "### Tree_Node\n",
    "\n",
    "* nodes in a motif tree represent lmer sequences which are at most 2d away from all other nodes in the path to the root <p></p>\n",
    "* contains lmer sequence, position, input sequence origin, depth, children, and parent <p></p>\n",
    "* child nodes can be added and removed <p></p>"
   ],
   "id": "323ccd540998370d"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-04-21T02:06:01.088730Z",
     "start_time": "2024-04-21T02:06:01.084714Z"
    }
   },
   "cell_type": "code",
   "source": [
    "class Tree_Node:\n",
    "    def __init__(self, seq, seq_id, pos, depth=0):\n",
    "        self.seq = seq\n",
    "        self.seq_id = seq_id\n",
    "        self.pos = pos\n",
    "        self.depth = depth\n",
    "        \n",
    "        self.children = []\n",
    "        self.parent = None\n",
    "\n",
    "    def add_child(self, child_node):\n",
    "        child_node.parent = self\n",
    "        child_node.depth = self.depth + 1\n",
    "        self.children.append(child_node)\n",
    "        \n",
    "\n",
    "    def remove_child(self, child_node):\n",
    "        if child_node in self.children:\n",
    "            child_node.parent = None\n",
    "            self.children.remove(child_node)"
   ],
   "id": "b912f4314f434eba",
   "outputs": [],
   "execution_count": 5
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "### Tree\n",
    "\n",
    "* constructed trees represent one or more cliques of motifs <p></p>\n",
    "* one path from root to leaf represents a clique of motif instances <p></p>\n",
    "* trees with multiple paths contains multiple cliques <p></p><p></p>\n",
    "* the root node is a tree node with lmer at most 2d from a reference pair <p></p>\n",
    "* all root nodes are from sequence 2 (reference pairs are from sequences 0 and 1) <p></p>\n",
    "* each level of the tree represents nodes from each input sequence (besides the ref seqs) <p></p>\n",
    "* leaves are stored for to be easily checked during branch pruning <p></p>"
   ],
   "id": "bb13ffe90217d8cc"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-04-21T02:06:01.851624Z",
     "start_time": "2024-04-21T02:06:01.848299Z"
    }
   },
   "cell_type": "code",
   "source": [
    "class Tree:\n",
    "    def __init__(self, root):\n",
    "        self.root = root\n",
    "        self.current_leaves = []\n",
    "        self.height = 0\n",
    "        self.id = 0\n",
    "    \n",
    "    def save_leaf(self, leaf):\n",
    "        self.current_leaves.append(leaf)\n",
    "        \n",
    "    def unsave_leaf(self, leaf):\n",
    "        if leaf in self.current_leaves:\n",
    "            self.current_leaves.remove(leaf)"
   ],
   "id": "53897b64ff3bdc9f",
   "outputs": [],
   "execution_count": 6
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "### Clique\n",
    "\n",
    "* a set of instances of a motif which are all at most 2d different <p></p>\n",
    "* a clique is initially formed by finding a path through a tree from root to leaf <p></p>\n",
    "* cliques can be merged if all lmers in one are at most 2d from all lmers in another <p></p>\n",
    "* merged cliques will then contain more instances than the number of input sequences <p></p>\n",
    "* for dedicated testing inputs where there is deliberately one motif per input sequence, merging may pick up random instances only similar by chance <p></p>\n",
    "* when number of occurrences of motif per input is unknown, this may help detect multiple motif instances in the same input sequence, but other modifications to the algorithm are required in that scenario <p></p>"
   ],
   "id": "4ebef81faf300d16"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-04-21T02:06:02.555865Z",
     "start_time": "2024-04-21T02:06:02.551754Z"
    }
   },
   "cell_type": "code",
   "source": [
    "class Clique:\n",
    "    def __init__(self, path):\n",
    "        self.nodes = path\n",
    "        self.seqs = []\n",
    "        for node in self.nodes:\n",
    "            self.seqs.append(node.seq)\n",
    "            \n",
    "        self.consensus = \"\"\n",
    "        self.score = -1\n",
    "        self.rank = -1\n",
    "        \n",
    "        self.merged = False\n",
    "        \n",
    "    # def __deepcopy__(self, memo):\n",
    "    #     new_nodes = []\n",
    "    #     new_seqs = []\n",
    "    #     for node in self.nodes:\n",
    "    #         new_nodes.append(Tree_Node(node.seq, node.seq_id, node.pos))\n",
    "    #         new_seqs.append(node.seq)\n",
    "    #     new_copy = Clique(new_nodes)\n",
    "    #     new_copy.seqs = new_seqs\n",
    "    #     return new_copy\n",
    "            "
   ],
   "id": "ba0bc128d8f6e0db",
   "outputs": [],
   "execution_count": 7
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "# FUNCTIONS #\n",
    "\n",
    "### Define subroutines to be used by the main steps of PyTreeMotif ###"
   ],
   "id": "8535d9cd69c1619e"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "### read_file_to_list \n",
    "\n",
    "* reads a text file of one input sequence per line into a python list of input sequences\n"
   ],
   "id": "21936dc4282f7ee6"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-04-21T02:06:06.295443Z",
     "start_time": "2024-04-21T02:06:06.291184Z"
    }
   },
   "cell_type": "code",
   "source": [
    "def read_file_to_list(filename):\n",
    "    try:\n",
    "        with open(filename, 'r') as file:\n",
    "            lines = file.readlines()\n",
    "        lines = [line.strip() for line in lines]\n",
    "        \n",
    "        return lines\n",
    "    except FileNotFoundError:\n",
    "        print(f\"Error: File '{filename}' not found.\")\n",
    "        return []\n",
    "    except Exception as e:\n",
    "        print(f\"Error: {e}\")\n",
    "        return []"
   ],
   "id": "8124253d90d9bbb",
   "outputs": [],
   "execution_count": 8
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "### hamming_distance\n",
    "\n",
    "* measures hamming distance, the number of single character differences, between two strings\n"
   ],
   "id": "59e321ba2e850a3"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-04-21T02:06:07.132412Z",
     "start_time": "2024-04-21T02:06:07.129777Z"
    }
   },
   "cell_type": "code",
   "source": [
    "def hamming_distance(str1, str2):\n",
    "\n",
    "    return sum(bit1 != bit2 for bit1, bit2 in zip(str1, str2))\n",
    "    "
   ],
   "id": "18d3bb406808d6c2",
   "outputs": [],
   "execution_count": 9
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "### extendable\n",
    "\n",
    "* given a tree and candidate node, finds all possible leaves to which the node could be added <p></p>\n",
    "* breadth-first traversal of the tree, adding all children that are similar to node to queue <p></p>\n",
    "* if the queue is empty, then the node could not be added to any leaves <p></p>\n",
    "* traversal ends after finishing second to lowest level <p></p>\n",
    "* all nodes left in the queue are the leaves of branches where every node in the branch all the way to the root is at most 2d from the new node <p></p>"
   ],
   "id": "7c456e9f1d68637d"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-04-21T02:06:08.309966Z",
     "start_time": "2024-04-21T02:06:08.304864Z"
    }
   },
   "cell_type": "code",
   "source": [
    "def extendable(node, tree, ith_seq, _2d):\n",
    "    queue = deque()\n",
    "    \n",
    "    # node must first be similar to root \n",
    "    if hamming_distance(tree.root.seq, node.seq) <= _2d:\n",
    "        queue.append(tree.root)\n",
    "        \n",
    "        # while there are still potential branches\n",
    "        while len(queue) > 0 and queue[0].depth < ith_seq - 3:\n",
    "            front = queue.popleft()\n",
    "            \n",
    "            # check if node is similar to any child nodes\n",
    "            for child in front.children:\n",
    "                if hamming_distance(child.seq, node.seq) <= _2d:\n",
    "                    queue.append(child)\n",
    "    return queue\n",
    "        "
   ],
   "id": "836ac43ace1fbaa8",
   "outputs": [],
   "execution_count": 10
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "### prune_branches\n",
    "\n",
    "* blah blah"
   ],
   "id": "6a6dbcc24711721e"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-04-21T02:06:11.120571Z",
     "start_time": "2024-04-21T02:06:11.116881Z"
    }
   },
   "cell_type": "code",
   "source": [
    "def prune_branches(tree):\n",
    "    for leaf in tree.current_leaves:\n",
    "        if leaf.depth < tree.height:\n",
    "            current_node = leaf\n",
    "            num_siblings = len(current_node.parent.children) - 1\n",
    "            branch_list = [leaf.seq]\n",
    "            while num_siblings == 0:\n",
    "                current_node = current_node.parent\n",
    "                num_siblings = len(current_node.parent.children) - 1\n",
    "                branch_list.append(current_node.seq)\n",
    "            parent = current_node.parent\n",
    "            parent.remove_child(current_node)\n",
    "            tree.unsave_leaf(leaf)\n",
    "            "
   ],
   "id": "8e24be551bba853b",
   "outputs": [],
   "execution_count": 11
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "### check_tree\n",
    "\n",
    "* blah blah"
   ],
   "id": "b88818dd92bb21e7"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-04-21T02:06:11.990425Z",
     "start_time": "2024-04-21T02:06:11.987519Z"
    }
   },
   "cell_type": "code",
   "source": [
    "def check_tree(tree):\n",
    "    is_valid = True\n",
    "    for leaf in tree.current_leaves:\n",
    "        if leaf.depth < tree.height:\n",
    "            is_valid = False\n",
    "            break\n",
    "    \n",
    "    return is_valid"
   ],
   "id": "ef8e5e598c7f53a1",
   "outputs": [],
   "execution_count": 12
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "### find_paths\n",
    "\n",
    "* blah blah"
   ],
   "id": "ae230a12f85c6727"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-04-21T02:06:12.926670Z",
     "start_time": "2024-04-21T02:06:12.917206Z"
    }
   },
   "cell_type": "code",
   "source": [
    "def find_paths(node, path, paths):\n",
    "    if node is None:\n",
    "        return\n",
    "    path.append(node)\n",
    "    if not node.children:\n",
    "        clique = Clique(path)\n",
    "        paths.append(clique)\n",
    "    \n",
    "    for child in node.children:\n",
    "        # branch_path = copy.deepcopy(path)\n",
    "        find_paths(child, path[:], paths)\n",
    "        \n",
    "    \n",
    "    \n",
    "    return paths"
   ],
   "id": "79bef1f6ef2fd760",
   "outputs": [],
   "execution_count": 13
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "### merge\n",
    "\n",
    "* UNDER CONSTRUCTION"
   ],
   "id": "b6925b489a1d2f64"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-04-21T02:06:13.450733Z",
     "start_time": "2024-04-21T02:06:13.446359Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# def merge(new_cliques, _2d):\n",
    "#     \n",
    "#     for prev_clique in merged_cliques:\n",
    "# \n",
    "#         match = True\n",
    "#         for seq_1 in prev_clique.seqs:\n",
    "#             if not match:\n",
    "#                 break\n",
    "#             for seq_2 in new_clique.seqs:\n",
    "#                 if hamming_distance(seq_1, seq_2) > _2d:\n",
    "#                     match = False\n",
    "#                     break\n",
    "# \n",
    "#         if match:\n",
    "#             merged_clique = copy.deepcopy(prev_clique)\n",
    "#             for node in new_clique.nodes:\n",
    "#                 if node.seq not in prev_clique.seqs:\n",
    "#                     copy_node = Tree_Node(node.seq, node.seq_id, node.pos)\n",
    "#                     merged_clique.nodes.append(copy_node)\n",
    "#                     merged_clique.seqs.append(node.seq)\n",
    "#             new_cliques.append(merged_clique)\n",
    "#     new_cliques.append(new_clique)\n",
    "#     merged_cliques.extend(new_cliques)"
   ],
   "id": "6c2753857434fd18",
   "outputs": [],
   "execution_count": 14
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "### score_motifs\n",
    "\n",
    "* blah blah"
   ],
   "id": "8140fa6c22ff03f"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-04-21T02:06:13.949808Z",
     "start_time": "2024-04-21T02:06:13.939898Z"
    }
   },
   "cell_type": "code",
   "source": [
    "def score_motifs(motifs, motif_length):\n",
    "    scores = []\n",
    "    for motif in motifs:\n",
    "        \n",
    "        profile = [[0] * motif_length for _ in range(4)]\n",
    "        indices = {\n",
    "            'A': 0,\n",
    "            'C': 1,\n",
    "            'G': 2,\n",
    "            'T': 3,\n",
    "            0: 'A',\n",
    "            1: 'C',\n",
    "            2: 'G',\n",
    "            3: 'T'\n",
    "        }\n",
    "        for i in range(len(motif.seqs)):\n",
    "            for j in range(motif_length):\n",
    "                profile[indices[motif.seqs[i][j]]][j] += 1\n",
    "                \n",
    "        consensus = \"\"\n",
    "                \n",
    "        for j in range(motif_length):\n",
    "            max_base = max(row[j] for row in profile)\n",
    "            \n",
    "            max_base_index = next(idx for idx, row in enumerate(profile) if row[j] == max_base)\n",
    "            consensus += indices[max_base_index]\n",
    "            \n",
    "        score = 0\n",
    "        for seq in motif.seqs:\n",
    "            score += hamming_distance(seq, consensus)\n",
    "        \n",
    "        motif.consensus = consensus\n",
    "        motif.score = score\n",
    "        scores.append((score, motif))\n",
    "    \n",
    "    sorted_scores = sorted(scores, key=lambda x: x[0])\n",
    "    \n",
    "    rank = 1\n",
    "    sorted_scores[0][1].rank = rank\n",
    "\n",
    "    for i in range(1, len(sorted_scores)):\n",
    "        if sorted_scores[i][0] > sorted_scores[i - 1][0]:\n",
    "            rank += 1\n",
    "        sorted_scores[i][1].rank = rank        \n",
    "            \n",
    "        "
   ],
   "id": "4d80601b59e72ffa",
   "outputs": [],
   "execution_count": 15
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "### consensus_grouping\n",
    "\n",
    "* blah blah"
   ],
   "id": "6608e648d53d6ce"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-04-21T02:06:14.479093Z",
     "start_time": "2024-04-21T02:06:14.471853Z"
    }
   },
   "cell_type": "code",
   "source": [
    "def consensus_grouping(motifs):\n",
    "    consensus_groups = {}\n",
    "    for motif in motifs:\n",
    "        if motif.consensus not in consensus_groups:\n",
    "            consensus_groups[motif.consensus] = [motif]\n",
    "        else:\n",
    "            consensus_groups[motif.consensus].append(motif)\n",
    "            \n",
    "    consensus_score_report = {}\n",
    "    \n",
    "    for consensus, motif_list in consensus_groups.items():\n",
    "        avg_score = sum(motif.score for motif in motif_list) / len(motif_list)\n",
    "        best_score = min(motif.score for motif in motif_list)\n",
    "        \n",
    "        consensus_score_report[consensus] = [best_score, avg_score]\n",
    "        \n",
    "    sorted_score_report = sorted(consensus_score_report.items(), key=lambda x: x[1][0])\n",
    "    ssr_dict = {consensus: scores for consensus, scores in sorted_score_report}\n",
    "    \n",
    "    return consensus_groups, ssr_dict"
   ],
   "id": "e526929faf2040e2",
   "outputs": [],
   "execution_count": 16
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "# STEP 1: NODE SELECTION #\n",
    "\n",
    "### Find reference pairs from the reference sequences. \n",
    "\n",
    "### Then, find lmers similar to the reference pair from all other input sequences\n",
    "\n",
    "### If a reference pair can't find at least one similar lmer from each input sequence, its not valid\n",
    "\n"
   ],
   "id": "cba84ca10ff13f44"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-04-21T02:06:15.030247Z",
     "start_time": "2024-04-21T02:06:15.023804Z"
    }
   },
   "cell_type": "code",
   "source": [
    "def select_nodes(sequence_list, l, _2d, show_progress):\n",
    "        \n",
    "    num_seq = len(sequence_list)\n",
    "        \n",
    "    reference_seq_1 = sequence_list[0]\n",
    "    reference_seq_2 = sequence_list[1]\n",
    "    reference_pair_set = set()\n",
    "    reference_pair_list = []\n",
    "    reference_pair_counter = 0\n",
    "        \n",
    "    for i in (tqdm(range(len(reference_seq_1) - l + 1), desc='Finding reference pairs and sequence nodes', position=0, leave=True, file=sys.stdout) if show_progress else range(len(reference_seq_1) - l + 1)):\n",
    "        for j in range(len(reference_seq_2) - l + 1):\n",
    "            if hamming_distance(reference_seq_1[i:i+l], reference_seq_2[j:j+l]) <= _2d:\n",
    "                ref_sub_seq = reference_seq_1[i:i+l]\n",
    "                ref_sub_seq2 = reference_seq_2[j:j+l]\n",
    "                if (ref_sub_seq, ref_sub_seq2) in reference_pair_set:\n",
    "                    continue\n",
    "                reference_pair = Reference_Pair(pair_id=reference_pair_counter, seq1=ref_sub_seq, seq2=ref_sub_seq2, pos1=i, pos2=j, num_seq=num_seq)    \n",
    "                \n",
    "                count_seqs_with_nodes = 0\n",
    "                for k in range(2, num_seq):                \n",
    "                    has_nodes = False\n",
    "                    for m in range(len(sequence_list[k]) - l + 1):\n",
    "                        sub_seq = sequence_list[k][m:m+l]\n",
    "                        if hamming_distance(sub_seq, ref_sub_seq) <= _2d and hamming_distance(sub_seq, ref_sub_seq2) <= _2d:\n",
    "                            seq_node = Sequence_Node(seq=sub_seq, id=k, pos=m)\n",
    "                            reference_pair.node_sets[k-2].add_member(seq_node)\n",
    "                            has_nodes = True\n",
    "                    if has_nodes:\n",
    "                        count_seqs_with_nodes += 1\n",
    "                \n",
    "                if count_seqs_with_nodes == num_seq - 2:\n",
    "                    reference_pair_list.append(reference_pair)\n",
    "                    reference_pair_set.add((ref_sub_seq, ref_sub_seq2))\n",
    "                    reference_pair_counter += 1\n",
    "                    \n",
    "    return reference_pair_list\n",
    "                \n",
    "\n",
    "            \n",
    "\n",
    "\n"
   ],
   "id": "bfcb5d26f7f910e1",
   "outputs": [],
   "execution_count": 17
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "# STEP 2: TREE CONSTRUCTION #",
   "id": "5c88dce40133a98b"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-04-21T02:06:16.130088Z",
     "start_time": "2024-04-21T02:06:16.122445Z"
    }
   },
   "cell_type": "code",
   "source": [
    "\n",
    "def construct_trees(reference_pair_list, _2d, show_progress):\n",
    "    \n",
    "    num_seq = len(reference_pair_list[0].node_sets) + 2\n",
    "    tree_count = 0\n",
    "    cliques = []\n",
    "    for pair in (tqdm(reference_pair_list, desc='Building trees from reference pairs', position=0, leave=True, file=sys.stdout) if show_progress else reference_pair_list):\n",
    "        node_sets = pair.node_sets\n",
    "        for root in node_sets[0].members:\n",
    "            root_node = Tree_Node(seq=root.seq, seq_id=root.id, pos=root.pos)\n",
    "            tree = Tree(root_node)\n",
    "            \n",
    "            for i in range(3, num_seq):\n",
    "                flag = False\n",
    "                for node in node_sets[i - 2].members:\n",
    "                    branches = extendable(node, tree, i, _2d)\n",
    "                    if branches:\n",
    "                        # print(\"branches check\")\n",
    "                        for branch in branches:\n",
    "                            new_leaf = Tree_Node(seq=node.seq, seq_id=node.id, pos=node.pos)\n",
    "                            branch.add_child(new_leaf)\n",
    "                            tree.save_leaf(new_leaf)\n",
    "                            tree.unsave_leaf(branch)\n",
    "                        flag = True\n",
    "                if not flag:\n",
    "                    tree = None\n",
    "                    break\n",
    "                tree.height += 1\n",
    "                prune_branches(tree)\n",
    "            if tree and check_tree(tree):    \n",
    "                pair.tree_list.append(tree)\n",
    "                tree.id = tree_count\n",
    "                tree_count += 1\n",
    "                \n",
    "                new_cliques = find_paths(tree.root, [], [])\n",
    "\n",
    "                for new_clique in new_cliques:\n",
    "                    \n",
    "                    new_clique.nodes.insert(0, Tree_Node(pair.seq2, 1, pair.pos2))\n",
    "                    new_clique.seqs.insert(0, pair.seq2)\n",
    "                    \n",
    "                    new_clique.nodes.insert(0, Tree_Node(pair.seq1, 0, pair.pos1))\n",
    "                    new_clique.seqs.insert(0, pair.seq1)\n",
    "                    \n",
    "                    cliques.append(new_clique)\n",
    "                    \n",
    "                    \n",
    "    return cliques\n"
   ],
   "id": "94eaeac5108e9493",
   "outputs": [],
   "execution_count": 18
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "<span style=\"font-size: 5em;\">PyTreeMotif</span>\n",
    "<p></p>\n",
    "\n",
    "### Define the full algorithm as the function find_motif() ###"
   ],
   "id": "11bc2b44e76196de"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-04-21T02:06:17.017468Z",
     "start_time": "2024-04-21T02:06:17.006891Z"
    }
   },
   "cell_type": "code",
   "source": [
    "def find_motif(DNA, motif_length, d=-1, print_output=True, show_progress=True, output_consensus_groups=False):\n",
    "    \n",
    "    DNA = [seq.upper() for seq in DNA]\n",
    "    reference_pairs = None\n",
    "    motifs = None\n",
    "    _2d = 0\n",
    "    \n",
    "    if d == -1:\n",
    "        i = 0\n",
    "        while not reference_pairs and not motifs and i < 5:\n",
    "            _2d = i * 2\n",
    "            reference_pairs = select_nodes(DNA, motif_length, _2d, show_progress)\n",
    "            if not reference_pairs:\n",
    "                i += 1\n",
    "                continue\n",
    "            motifs = construct_trees(reference_pairs, _2d, show_progress)\n",
    "            i += 1\n",
    "    else:\n",
    "        _2d = d * 2\n",
    "        reference_pairs = select_nodes(DNA, motif_length, _2d, show_progress)\n",
    "        motifs = construct_trees(reference_pairs, _2d, show_progress)\n",
    "    \n",
    "    if not reference_pairs or not motifs:\n",
    "        return [\"no motifs found :(\"]\n",
    "    \n",
    "    score_motifs(motifs, motif_length)\n",
    "    score_sorted_motifs = sorted(motifs, key=lambda motif: motif.rank)\n",
    "    \n",
    "    best_motifs = []\n",
    "    i = 0\n",
    "    while i < len(score_sorted_motifs) and i < 3:\n",
    "        best_motifs.append(score_sorted_motifs[i])\n",
    "        i += 1\n",
    "        \n",
    "    consensus_groups, sorted_group_score_report = consensus_grouping(score_sorted_motifs)\n",
    "    \n",
    "    if print_output:\n",
    "        \n",
    "        dash_line = '\\n' + ('-' * 80) + '\\n'\n",
    "        \n",
    "        if output_consensus_groups:\n",
    "            print(\"\\n\\n ### CONSENSUS GROUPS ### \")\n",
    "            print(dash_line)\n",
    "            for consensus, scores in sorted_group_score_report.items():\n",
    "                print(\"Consensus: \" + consensus)\n",
    "                print(\"Number of cliques: \" + str(len(consensus_groups[consensus])))\n",
    "                print(\"Best score of this group: \" + str(scores[0]))\n",
    "                print(\"Average score of this group: \" + str(scores[1]))\n",
    "                print(\"Motifs:\")\n",
    "                for motif in consensus_groups[consensus]:\n",
    "                    print(\"Rank: \" + str(motif.rank) + \", Score: \" + str(motif.score))\n",
    "                    for instance in motif.nodes:\n",
    "                        print(\"Seq: \" + str(instance.seq_id) + \", Position: \" + str(instance.pos) + \", \" + str(instance.seq))\n",
    "                    print()\n",
    "                print(dash_line)\n",
    "\n",
    "                \n",
    "\n",
    "        print(\"Total Cliques: \" + str(len(motifs)))\n",
    "        print(\"Total Consensus Motifs: \" + str(len(consensus_groups)))\n",
    "        if d == -1:\n",
    "            print(\"Substitutions Allowed: \" + str(_2d // 2))\n",
    "        print(dash_line) \n",
    "        \n",
    "        print(\"Top 3 Motifs\")\n",
    "        print(dash_line)\n",
    "        for winner in best_motifs:\n",
    "            print(\"Consensus: \" + str(winner.consensus) + \", Rank: \" + str(winner.rank) + \", Score: \" + str(winner.score))\n",
    "            print(\"Instances: \")\n",
    "            for node in winner.nodes:\n",
    "                print(\"Seq: \" + str(node.seq_id) + \", Position: \" + str(node.pos) + \", \" + str(node.seq))\n",
    "            print()\n",
    "            \n",
    "            \n",
    "           \n",
    "    \n",
    "    return [motif.consensus for motif in best_motifs]\n",
    "    "
   ],
   "id": "73b8b68a6f3af8c3",
   "outputs": [],
   "execution_count": 19
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "# TEST CASES #",
   "id": "134bad2d2699f752"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "### First, test with the simple test cases provided on the course website ###",
   "id": "bc4db6b3179ea77c"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-04-21T02:06:18.567842Z",
     "start_time": "2024-04-21T02:06:18.528691Z"
    }
   },
   "cell_type": "code",
   "source": [
    "test1 = read_file_to_list(\"data1.txt\")\n",
    "\n",
    "find_motif(test1, 10)"
   ],
   "id": "c1866d602f7c136",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Finding reference pairs and sequence nodes: 100%|██████████| 91/91 [00:00<00:00, 4042.38it/s]\n",
      "Building trees from reference pairs: 100%|██████████| 1/1 [00:00<00:00, 4284.27it/s]\n",
      "Total Cliques: 1\n",
      "Total Consensus Motifs: 1\n",
      "Substitutions Allowed: 0\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "\n",
      "Top 3 Motifs\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "\n",
      "Consensus: GGGTCTAAGC, Rank: 1, Score: 0\n",
      "Instances: \n",
      "Seq: 0, Position: 67, GGGTCTAAGC\n",
      "Seq: 1, Position: 34, GGGTCTAAGC\n",
      "Seq: 2, Position: 12, GGGTCTAAGC\n",
      "Seq: 3, Position: 8, GGGTCTAAGC\n",
      "Seq: 4, Position: 14, GGGTCTAAGC\n",
      "Seq: 5, Position: 90, GGGTCTAAGC\n",
      "Seq: 6, Position: 51, GGGTCTAAGC\n",
      "Seq: 7, Position: 75, GGGTCTAAGC\n",
      "Seq: 8, Position: 56, GGGTCTAAGC\n",
      "Seq: 9, Position: 83, GGGTCTAAGC\n",
      "Seq: 10, Position: 72, GGGTCTAAGC\n",
      "Seq: 11, Position: 55, GGGTCTAAGC\n",
      "Seq: 12, Position: 34, GGGTCTAAGC\n",
      "Seq: 13, Position: 62, GGGTCTAAGC\n",
      "Seq: 14, Position: 44, GGGTCTAAGC\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['GGGTCTAAGC']"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 20
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-04-21T02:06:19.187992Z",
     "start_time": "2024-04-21T02:06:19.139258Z"
    }
   },
   "cell_type": "code",
   "source": [
    "test2 = read_file_to_list(\"data2.txt\")\n",
    "\n",
    "find_motif(test2, 10)"
   ],
   "id": "b6a9e0197ced2ac3",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Finding reference pairs and sequence nodes: 100%|██████████| 91/91 [00:00<00:00, 4581.13it/s]\n",
      "Finding reference pairs and sequence nodes: 100%|██████████| 91/91 [00:00<00:00, 4816.11it/s]\n",
      "Building trees from reference pairs: 100%|██████████| 1/1 [00:00<00:00, 4782.56it/s]\n",
      "Total Cliques: 1\n",
      "Total Consensus Motifs: 1\n",
      "Substitutions Allowed: 1\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "\n",
      "Top 3 Motifs\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "\n",
      "Consensus: GGGTCTAAGC, Rank: 1, Score: 14\n",
      "Instances: \n",
      "Seq: 0, Position: 47, GGGTCTAAGG\n",
      "Seq: 1, Position: 41, GGGCCTAAGC\n",
      "Seq: 2, Position: 41, GGATCTAAGC\n",
      "Seq: 3, Position: 42, GGGTCTAACC\n",
      "Seq: 4, Position: 17, GGGTCTAAGC\n",
      "Seq: 5, Position: 60, GGGTCTAGGC\n",
      "Seq: 6, Position: 37, GGGTCGAAGC\n",
      "Seq: 7, Position: 18, GGGTCTAATC\n",
      "Seq: 8, Position: 43, GGGTCTGAGC\n",
      "Seq: 9, Position: 17, GGATCTAAGC\n",
      "Seq: 10, Position: 8, GGGTCAAAGC\n",
      "Seq: 11, Position: 83, AGGTCTAAGC\n",
      "Seq: 12, Position: 33, GGGTCGAAGC\n",
      "Seq: 13, Position: 20, GGGGCTAAGC\n",
      "Seq: 14, Position: 84, GGGTCCAAGC\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['GGGTCTAAGC']"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 21
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "# 15,4 BENCHMARKING #\n",
    "### Test the algorithm with the original subtle motif challenge proposed by Pevzner ###\n",
    "\n",
    "First define the function to generate a single dataset.\n",
    "\n",
    "It will take a sequence length argument.\n",
    "\n",
    "First, a 15mer motif is randomly generated.\n",
    "\n",
    "For each of 20 sequences:\n",
    "\n",
    "1. The sequence is randomly generated at the given length\n",
    "    \n",
    "2. The motif is mutated 4 times at random positions with random bases\n",
    "    \n",
    "3. The mutated motif is implanted into the sequence at a random position\n",
    "\n",
    "The dataset is now made up of 20 random sequences of the given length, each with a (15,4) implanted motif\n"
   ],
   "id": "1b0e8f65e16c8074"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-04-21T02:06:22.048117Z",
     "start_time": "2024-04-21T02:06:22.041238Z"
    }
   },
   "cell_type": "code",
   "source": [
    "def generate_15_4_dataset(seq_len, l, d):\n",
    "    bases = ['A', 'C', 'G', 'T']\n",
    "    \n",
    "    motif = ''.join(random.choice(bases) for _ in range(l))\n",
    "    \n",
    "    # print(\"Implanted motif: \" + motif)\n",
    "    # print()\n",
    "    \n",
    "    dataset = []\n",
    "    implanted_motifs = []\n",
    "    \n",
    "    for i in range(20):\n",
    "        input_seq = ''.join(random.choice(bases) for _ in range(seq_len))\n",
    "        \n",
    "        sub_positions = [random.randint(0, l-1) for _ in range(d)]\n",
    "        \n",
    "        implanted_motif = motif\n",
    "        \n",
    "        for position in sub_positions:\n",
    "            mutation = random.choice(bases)\n",
    "            implanted_motif = implanted_motif[:position] + mutation + implanted_motif[position + 1:]\n",
    "            \n",
    "        implant_position = random.randint(0, seq_len - l)\n",
    "        \n",
    "        implanted_motifs.append((implanted_motif, implant_position))\n",
    "            \n",
    "        input_seq_w_motif = input_seq[:implant_position] + implanted_motif + input_seq[implant_position + l:]\n",
    "        \n",
    "        dataset.append(input_seq_w_motif)\n",
    "    \n",
    "    # for im in implanted_motifs:\n",
    "    #     print(im[0], im[1])\n",
    "        \n",
    "    return dataset, motif, implanted_motifs\n",
    "        \n"
   ],
   "id": "c3991ad1c79367c1",
   "outputs": [],
   "execution_count": 22
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "### Test the 15,4 dataset generator ###",
   "id": "155a2df4c546c0a3"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-04-21T02:06:23.023832Z",
     "start_time": "2024-04-21T02:06:23.018695Z"
    }
   },
   "cell_type": "code",
   "source": [
    "dataset, implanted_motif, instances = generate_15_4_dataset(200, 15, 4)\n",
    "print(\"Implanted Motif: \" + implanted_motif)\n",
    "print(\"Instances: \")\n",
    "for instance in instances:\n",
    "    print(instance)"
   ],
   "id": "42a89cf395f1de78",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Implanted Motif: CGTAACGATCGAATC\n",
      "Instances: \n",
      "('CGGAACGTTCGAAAC', 149)\n",
      "('CAAAACGCTCAAATC', 180)\n",
      "('CGTAATGATCGACTC', 24)\n",
      "('CGTAACGAGTGAGAC', 171)\n",
      "('TGTAACTATCAAGTC', 183)\n",
      "('CGTGACGATTGAATC', 69)\n",
      "('CGCGACGTTCGAATC', 89)\n",
      "('TGGTACGATCGAATC', 130)\n",
      "('CGGAACGATCGGATC', 70)\n",
      "('TGTAACCATCGGATC', 123)\n",
      "('CGTAAGTATCGAATC', 49)\n",
      "('AGTAACCATCGAATC', 40)\n",
      "('CAGCACGATCGCATC', 115)\n",
      "('CGTAAAGATCGAATC', 49)\n",
      "('CTTAACGCTCGAATC', 49)\n",
      "('CGTGCCGATCGAATC', 116)\n",
      "('CGTAACGTGCGAATC', 78)\n",
      "('CGTAGCGATCGAAAC', 158)\n",
      "('CTTATCGATCGAACG', 56)\n",
      "('TGTAACGATCGAAGC', 27)\n"
     ]
    }
   ],
   "execution_count": 23
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "### Test the algorithm on the above 15,4 dataset ###\n",
    "\n",
    "while we're here, test outputting consensus groups too"
   ],
   "id": "7c8585c8286b027b"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-04-21T02:06:35.491600Z",
     "start_time": "2024-04-21T02:06:24.184133Z"
    }
   },
   "cell_type": "code",
   "source": "found_motif = find_motif(dataset, 15, 4, output_consensus_groups=True)\n",
   "id": "c2d8bc1b0a19724d",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Finding reference pairs and sequence nodes: 100%|██████████| 186/186 [00:11<00:00, 16.51it/s]\n",
      "Building trees from reference pairs: 100%|██████████| 148/148 [00:00<00:00, 4958.08it/s]\n",
      "\n",
      "\n",
      " ### CONSENSUS GROUPS ### \n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "\n",
      "Consensus: CGTAACGATCGAATC\n",
      "Number of cliques: 6\n",
      "Best score of this group: 53\n",
      "Average score of this group: 56.0\n",
      "Motifs:\n",
      "Rank: 1, Score: 53\n",
      "Seq: 0, Position: 149, CGGAACGTTCGAAAC\n",
      "Seq: 1, Position: 180, CAAAACGCTCAAATC\n",
      "Seq: 2, Position: 24, CGTAATGATCGACTC\n",
      "Seq: 3, Position: 171, CGTAACGAGTGAGAC\n",
      "Seq: 4, Position: 183, TGTAACTATCAAGTC\n",
      "Seq: 5, Position: 69, CGTGACGATTGAATC\n",
      "Seq: 6, Position: 89, CGCGACGTTCGAATC\n",
      "Seq: 7, Position: 130, TGGTACGATCGAATC\n",
      "Seq: 8, Position: 70, CGGAACGATCGGATC\n",
      "Seq: 9, Position: 123, TGTAACCATCGGATC\n",
      "Seq: 10, Position: 49, CGTAAGTATCGAATC\n",
      "Seq: 11, Position: 40, AGTAACCATCGAATC\n",
      "Seq: 12, Position: 115, CAGCACGATCGCATC\n",
      "Seq: 13, Position: 49, CGTAAAGATCGAATC\n",
      "Seq: 14, Position: 49, CTTAACGCTCGAATC\n",
      "Seq: 15, Position: 116, CGTGCCGATCGAATC\n",
      "Seq: 16, Position: 78, CGTAACGTGCGAATC\n",
      "Seq: 17, Position: 158, CGTAGCGATCGAAAC\n",
      "Seq: 18, Position: 56, CTTATCGATCGAACG\n",
      "Seq: 19, Position: 27, TGTAACGATCGAAGC\n",
      "\n",
      "Rank: 2, Score: 55\n",
      "Seq: 0, Position: 149, CGGAACGTTCGAAAC\n",
      "Seq: 1, Position: 180, CAAAACGCTCAAATC\n",
      "Seq: 2, Position: 24, CGTAATGATCGACTC\n",
      "Seq: 3, Position: 171, CGTAACGAGTGAGAC\n",
      "Seq: 4, Position: 183, TGTAACTATCAAGTC\n",
      "Seq: 5, Position: 69, CGTGACGATTGAATC\n",
      "Seq: 6, Position: 89, CGCGACGTTCGAATC\n",
      "Seq: 7, Position: 139, CGAATCAATCGCGTC\n",
      "Seq: 8, Position: 70, CGGAACGATCGGATC\n",
      "Seq: 9, Position: 123, TGTAACCATCGGATC\n",
      "Seq: 10, Position: 49, CGTAAGTATCGAATC\n",
      "Seq: 11, Position: 40, AGTAACCATCGAATC\n",
      "Seq: 12, Position: 115, CAGCACGATCGCATC\n",
      "Seq: 13, Position: 49, CGTAAAGATCGAATC\n",
      "Seq: 14, Position: 49, CTTAACGCTCGAATC\n",
      "Seq: 15, Position: 116, CGTGCCGATCGAATC\n",
      "Seq: 16, Position: 78, CGTAACGTGCGAATC\n",
      "Seq: 17, Position: 158, CGTAGCGATCGAAAC\n",
      "Seq: 18, Position: 56, CTTATCGATCGAACG\n",
      "Seq: 19, Position: 27, TGTAACGATCGAAGC\n",
      "\n",
      "Rank: 2, Score: 55\n",
      "Seq: 0, Position: 149, CGGAACGTTCGAAAC\n",
      "Seq: 1, Position: 180, CAAAACGCTCAAATC\n",
      "Seq: 2, Position: 24, CGTAATGATCGACTC\n",
      "Seq: 3, Position: 162, GCGAACTATCGTAAC\n",
      "Seq: 4, Position: 183, TGTAACTATCAAGTC\n",
      "Seq: 5, Position: 69, CGTGACGATTGAATC\n",
      "Seq: 6, Position: 89, CGCGACGTTCGAATC\n",
      "Seq: 7, Position: 130, TGGTACGATCGAATC\n",
      "Seq: 8, Position: 70, CGGAACGATCGGATC\n",
      "Seq: 9, Position: 123, TGTAACCATCGGATC\n",
      "Seq: 10, Position: 49, CGTAAGTATCGAATC\n",
      "Seq: 11, Position: 40, AGTAACCATCGAATC\n",
      "Seq: 12, Position: 115, CAGCACGATCGCATC\n",
      "Seq: 13, Position: 49, CGTAAAGATCGAATC\n",
      "Seq: 14, Position: 49, CTTAACGCTCGAATC\n",
      "Seq: 15, Position: 116, CGTGCCGATCGAATC\n",
      "Seq: 16, Position: 78, CGTAACGTGCGAATC\n",
      "Seq: 17, Position: 158, CGTAGCGATCGAAAC\n",
      "Seq: 18, Position: 56, CTTATCGATCGAACG\n",
      "Seq: 19, Position: 27, TGTAACGATCGAAGC\n",
      "\n",
      "Rank: 3, Score: 57\n",
      "Seq: 0, Position: 149, CGGAACGTTCGAAAC\n",
      "Seq: 1, Position: 180, CAAAACGCTCAAATC\n",
      "Seq: 2, Position: 24, CGTAATGATCGACTC\n",
      "Seq: 3, Position: 171, CGTAACGAGTGAGAC\n",
      "Seq: 4, Position: 183, TGTAACTATCAAGTC\n",
      "Seq: 5, Position: 69, CGTGACGATTGAATC\n",
      "Seq: 6, Position: 89, CGCGACGTTCGAATC\n",
      "Seq: 7, Position: 130, TGGTACGATCGAATC\n",
      "Seq: 8, Position: 70, CGGAACGATCGGATC\n",
      "Seq: 9, Position: 123, TGTAACCATCGGATC\n",
      "Seq: 10, Position: 49, CGTAAGTATCGAATC\n",
      "Seq: 11, Position: 40, AGTAACCATCGAATC\n",
      "Seq: 12, Position: 115, CAGCACGATCGCATC\n",
      "Seq: 13, Position: 49, CGTAAAGATCGAATC\n",
      "Seq: 14, Position: 49, CTTAACGCTCGAATC\n",
      "Seq: 15, Position: 116, CGTGCCGATCGAATC\n",
      "Seq: 16, Position: 78, CGTAACGTGCGAATC\n",
      "Seq: 17, Position: 167, CGAAACAATTTGATG\n",
      "Seq: 18, Position: 56, CTTATCGATCGAACG\n",
      "Seq: 19, Position: 27, TGTAACGATCGAAGC\n",
      "\n",
      "Rank: 3, Score: 57\n",
      "Seq: 0, Position: 149, CGGAACGTTCGAAAC\n",
      "Seq: 1, Position: 180, CAAAACGCTCAAATC\n",
      "Seq: 2, Position: 24, CGTAATGATCGACTC\n",
      "Seq: 3, Position: 162, GCGAACTATCGTAAC\n",
      "Seq: 4, Position: 183, TGTAACTATCAAGTC\n",
      "Seq: 5, Position: 69, CGTGACGATTGAATC\n",
      "Seq: 6, Position: 89, CGCGACGTTCGAATC\n",
      "Seq: 7, Position: 139, CGAATCAATCGCGTC\n",
      "Seq: 8, Position: 70, CGGAACGATCGGATC\n",
      "Seq: 9, Position: 123, TGTAACCATCGGATC\n",
      "Seq: 10, Position: 49, CGTAAGTATCGAATC\n",
      "Seq: 11, Position: 40, AGTAACCATCGAATC\n",
      "Seq: 12, Position: 115, CAGCACGATCGCATC\n",
      "Seq: 13, Position: 49, CGTAAAGATCGAATC\n",
      "Seq: 14, Position: 49, CTTAACGCTCGAATC\n",
      "Seq: 15, Position: 116, CGTGCCGATCGAATC\n",
      "Seq: 16, Position: 78, CGTAACGTGCGAATC\n",
      "Seq: 17, Position: 158, CGTAGCGATCGAAAC\n",
      "Seq: 18, Position: 56, CTTATCGATCGAACG\n",
      "Seq: 19, Position: 27, TGTAACGATCGAAGC\n",
      "\n",
      "Rank: 4, Score: 59\n",
      "Seq: 0, Position: 149, CGGAACGTTCGAAAC\n",
      "Seq: 1, Position: 180, CAAAACGCTCAAATC\n",
      "Seq: 2, Position: 24, CGTAATGATCGACTC\n",
      "Seq: 3, Position: 171, CGTAACGAGTGAGAC\n",
      "Seq: 4, Position: 183, TGTAACTATCAAGTC\n",
      "Seq: 5, Position: 69, CGTGACGATTGAATC\n",
      "Seq: 6, Position: 89, CGCGACGTTCGAATC\n",
      "Seq: 7, Position: 139, CGAATCAATCGCGTC\n",
      "Seq: 8, Position: 70, CGGAACGATCGGATC\n",
      "Seq: 9, Position: 123, TGTAACCATCGGATC\n",
      "Seq: 10, Position: 49, CGTAAGTATCGAATC\n",
      "Seq: 11, Position: 40, AGTAACCATCGAATC\n",
      "Seq: 12, Position: 115, CAGCACGATCGCATC\n",
      "Seq: 13, Position: 49, CGTAAAGATCGAATC\n",
      "Seq: 14, Position: 49, CTTAACGCTCGAATC\n",
      "Seq: 15, Position: 116, CGTGCCGATCGAATC\n",
      "Seq: 16, Position: 78, CGTAACGTGCGAATC\n",
      "Seq: 17, Position: 167, CGAAACAATTTGATG\n",
      "Seq: 18, Position: 56, CTTATCGATCGAACG\n",
      "Seq: 19, Position: 27, TGTAACGATCGAAGC\n",
      "\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "\n",
      "Consensus: GTAACGATCGAATCC\n",
      "Number of cliques: 2\n",
      "Best score of this group: 62\n",
      "Average score of this group: 63.5\n",
      "Motifs:\n",
      "Rank: 5, Score: 62\n",
      "Seq: 0, Position: 150, GGAACGTTCGAAACA\n",
      "Seq: 1, Position: 181, AAAACGCTCAAATCC\n",
      "Seq: 2, Position: 25, GTAATGATCGACTCC\n",
      "Seq: 3, Position: 163, CGAACTATCGTAACG\n",
      "Seq: 4, Position: 184, GTAACTATCAAGTCC\n",
      "Seq: 5, Position: 70, GTGACGATTGAATCA\n",
      "Seq: 6, Position: 90, GCGACGTTCGAATCT\n",
      "Seq: 7, Position: 131, GGTACGATCGAATCA\n",
      "Seq: 8, Position: 71, GGAACGATCGGATCG\n",
      "Seq: 9, Position: 124, GTAACCATCGGATCA\n",
      "Seq: 10, Position: 50, GTAAGTATCGAATCC\n",
      "Seq: 11, Position: 41, GTAACCATCGAATCC\n",
      "Seq: 12, Position: 116, AGCACGATCGCATCG\n",
      "Seq: 13, Position: 50, GTAAAGATCGAATCT\n",
      "Seq: 14, Position: 50, TTAACGCTCGAATCG\n",
      "Seq: 15, Position: 117, GTGCCGATCGAATCA\n",
      "Seq: 16, Position: 79, GTAACGTGCGAATCT\n",
      "Seq: 17, Position: 159, GTAGCGATCGAAACA\n",
      "Seq: 18, Position: 57, TTATCGATCGAACGC\n",
      "Seq: 19, Position: 28, GTAACGATCGAAGCC\n",
      "\n",
      "Rank: 6, Score: 65\n",
      "Seq: 0, Position: 150, GGAACGTTCGAAACA\n",
      "Seq: 1, Position: 181, AAAACGCTCAAATCC\n",
      "Seq: 2, Position: 25, GTAATGATCGACTCC\n",
      "Seq: 3, Position: 163, CGAACTATCGTAACG\n",
      "Seq: 4, Position: 184, GTAACTATCAAGTCC\n",
      "Seq: 5, Position: 70, GTGACGATTGAATCA\n",
      "Seq: 6, Position: 90, GCGACGTTCGAATCT\n",
      "Seq: 7, Position: 140, GAATCAATCGCGTCG\n",
      "Seq: 8, Position: 71, GGAACGATCGGATCG\n",
      "Seq: 9, Position: 124, GTAACCATCGGATCA\n",
      "Seq: 10, Position: 50, GTAAGTATCGAATCC\n",
      "Seq: 11, Position: 41, GTAACCATCGAATCC\n",
      "Seq: 12, Position: 116, AGCACGATCGCATCG\n",
      "Seq: 13, Position: 50, GTAAAGATCGAATCT\n",
      "Seq: 14, Position: 50, TTAACGCTCGAATCG\n",
      "Seq: 15, Position: 117, GTGCCGATCGAATCA\n",
      "Seq: 16, Position: 79, GTAACGTGCGAATCT\n",
      "Seq: 17, Position: 159, GTAGCGATCGAAACA\n",
      "Seq: 18, Position: 57, TTATCGATCGAACGC\n",
      "Seq: 19, Position: 28, GTAACGATCGAAGCC\n",
      "\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "\n",
      "Consensus: TAACGATCGAATCCC\n",
      "Number of cliques: 1\n",
      "Best score of this group: 70\n",
      "Average score of this group: 70.0\n",
      "Motifs:\n",
      "Rank: 7, Score: 70\n",
      "Seq: 0, Position: 151, GAACGTTCGAAACAC\n",
      "Seq: 1, Position: 182, AAACGCTCAAATCCG\n",
      "Seq: 2, Position: 26, TAATGATCGACTCCC\n",
      "Seq: 3, Position: 164, GAACTATCGTAACGA\n",
      "Seq: 4, Position: 185, TAACTATCAAGTCCT\n",
      "Seq: 5, Position: 71, TGACGATTGAATCAC\n",
      "Seq: 6, Position: 91, CGACGTTCGAATCTG\n",
      "Seq: 7, Position: 132, GTACGATCGAATCAA\n",
      "Seq: 8, Position: 72, GAACGATCGGATCGT\n",
      "Seq: 9, Position: 125, TAACCATCGGATCAT\n",
      "Seq: 10, Position: 51, TAAGTATCGAATCCT\n",
      "Seq: 11, Position: 42, TAACCATCGAATCCC\n",
      "Seq: 12, Position: 117, GCACGATCGCATCGC\n",
      "Seq: 13, Position: 51, TAAAGATCGAATCTC\n",
      "Seq: 14, Position: 51, TAACGCTCGAATCGG\n",
      "Seq: 15, Position: 118, TGCCGATCGAATCAC\n",
      "Seq: 16, Position: 80, TAACGTGCGAATCTT\n",
      "Seq: 17, Position: 160, TAGCGATCGAAACAA\n",
      "Seq: 18, Position: 58, TATCGATCGAACGCG\n",
      "Seq: 19, Position: 29, TAACGATCGAAGCCT\n",
      "\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "\n",
      "Total Cliques: 9\n",
      "Total Consensus Motifs: 3\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "\n",
      "Top 3 Motifs\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "\n",
      "Consensus: CGTAACGATCGAATC, Rank: 1, Score: 53\n",
      "Instances: \n",
      "Seq: 0, Position: 149, CGGAACGTTCGAAAC\n",
      "Seq: 1, Position: 180, CAAAACGCTCAAATC\n",
      "Seq: 2, Position: 24, CGTAATGATCGACTC\n",
      "Seq: 3, Position: 171, CGTAACGAGTGAGAC\n",
      "Seq: 4, Position: 183, TGTAACTATCAAGTC\n",
      "Seq: 5, Position: 69, CGTGACGATTGAATC\n",
      "Seq: 6, Position: 89, CGCGACGTTCGAATC\n",
      "Seq: 7, Position: 130, TGGTACGATCGAATC\n",
      "Seq: 8, Position: 70, CGGAACGATCGGATC\n",
      "Seq: 9, Position: 123, TGTAACCATCGGATC\n",
      "Seq: 10, Position: 49, CGTAAGTATCGAATC\n",
      "Seq: 11, Position: 40, AGTAACCATCGAATC\n",
      "Seq: 12, Position: 115, CAGCACGATCGCATC\n",
      "Seq: 13, Position: 49, CGTAAAGATCGAATC\n",
      "Seq: 14, Position: 49, CTTAACGCTCGAATC\n",
      "Seq: 15, Position: 116, CGTGCCGATCGAATC\n",
      "Seq: 16, Position: 78, CGTAACGTGCGAATC\n",
      "Seq: 17, Position: 158, CGTAGCGATCGAAAC\n",
      "Seq: 18, Position: 56, CTTATCGATCGAACG\n",
      "Seq: 19, Position: 27, TGTAACGATCGAAGC\n",
      "\n",
      "Consensus: CGTAACGATCGAATC, Rank: 2, Score: 55\n",
      "Instances: \n",
      "Seq: 0, Position: 149, CGGAACGTTCGAAAC\n",
      "Seq: 1, Position: 180, CAAAACGCTCAAATC\n",
      "Seq: 2, Position: 24, CGTAATGATCGACTC\n",
      "Seq: 3, Position: 171, CGTAACGAGTGAGAC\n",
      "Seq: 4, Position: 183, TGTAACTATCAAGTC\n",
      "Seq: 5, Position: 69, CGTGACGATTGAATC\n",
      "Seq: 6, Position: 89, CGCGACGTTCGAATC\n",
      "Seq: 7, Position: 139, CGAATCAATCGCGTC\n",
      "Seq: 8, Position: 70, CGGAACGATCGGATC\n",
      "Seq: 9, Position: 123, TGTAACCATCGGATC\n",
      "Seq: 10, Position: 49, CGTAAGTATCGAATC\n",
      "Seq: 11, Position: 40, AGTAACCATCGAATC\n",
      "Seq: 12, Position: 115, CAGCACGATCGCATC\n",
      "Seq: 13, Position: 49, CGTAAAGATCGAATC\n",
      "Seq: 14, Position: 49, CTTAACGCTCGAATC\n",
      "Seq: 15, Position: 116, CGTGCCGATCGAATC\n",
      "Seq: 16, Position: 78, CGTAACGTGCGAATC\n",
      "Seq: 17, Position: 158, CGTAGCGATCGAAAC\n",
      "Seq: 18, Position: 56, CTTATCGATCGAACG\n",
      "Seq: 19, Position: 27, TGTAACGATCGAAGC\n",
      "\n",
      "Consensus: CGTAACGATCGAATC, Rank: 2, Score: 55\n",
      "Instances: \n",
      "Seq: 0, Position: 149, CGGAACGTTCGAAAC\n",
      "Seq: 1, Position: 180, CAAAACGCTCAAATC\n",
      "Seq: 2, Position: 24, CGTAATGATCGACTC\n",
      "Seq: 3, Position: 162, GCGAACTATCGTAAC\n",
      "Seq: 4, Position: 183, TGTAACTATCAAGTC\n",
      "Seq: 5, Position: 69, CGTGACGATTGAATC\n",
      "Seq: 6, Position: 89, CGCGACGTTCGAATC\n",
      "Seq: 7, Position: 130, TGGTACGATCGAATC\n",
      "Seq: 8, Position: 70, CGGAACGATCGGATC\n",
      "Seq: 9, Position: 123, TGTAACCATCGGATC\n",
      "Seq: 10, Position: 49, CGTAAGTATCGAATC\n",
      "Seq: 11, Position: 40, AGTAACCATCGAATC\n",
      "Seq: 12, Position: 115, CAGCACGATCGCATC\n",
      "Seq: 13, Position: 49, CGTAAAGATCGAATC\n",
      "Seq: 14, Position: 49, CTTAACGCTCGAATC\n",
      "Seq: 15, Position: 116, CGTGCCGATCGAATC\n",
      "Seq: 16, Position: 78, CGTAACGTGCGAATC\n",
      "Seq: 17, Position: 158, CGTAGCGATCGAAAC\n",
      "Seq: 18, Position: 56, CTTATCGATCGAACG\n",
      "Seq: 19, Position: 27, TGTAACGATCGAAGC\n",
      "\n"
     ]
    }
   ],
   "execution_count": 24
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "### Check the results ###",
   "id": "e5012d1ccc7e84b6"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-04-21T02:06:36.865232Z",
     "start_time": "2024-04-21T02:06:36.861876Z"
    }
   },
   "cell_type": "code",
   "source": [
    "for best_motif in found_motif:\n",
    "    try:\n",
    "        assert best_motif == implanted_motif\n",
    "        print(\"Success\")\n",
    "    except AssertionError:\n",
    "        print(\"Fail\")\n",
    "        continue\n",
    "print(implanted_motif)\n",
    "print(len(found_motif))"
   ],
   "id": "626905a34d4361d1",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Success\n",
      "Success\n",
      "Success\n",
      "CGTAACGATCGAATC\n",
      "3\n"
     ]
    }
   ],
   "execution_count": 25
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "### Test with different l,d motifs",
   "id": "8f972cbc80eae4c"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-04-20T20:47:56.076496Z",
     "start_time": "2024-04-20T20:47:46.525267Z"
    }
   },
   "cell_type": "code",
   "source": [
    "dataset, implanted_motif, instances = generate_15_4_dataset(200, 12, 3)\n",
    "_12_3_found_motif = find_motif(dataset, 12, 3, output_consensus_groups=True)"
   ],
   "id": "3e326e6dfe7d8d11",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Finding reference pairs and sequence nodes: 100%|██████████| 189/189 [00:09<00:00, 19.85it/s]\n",
      "Building trees from reference pairs: 100%|██████████| 172/172 [00:00<00:00, 9016.40it/s]\n",
      "\n",
      "\n",
      " ### CONSENSUS GROUPS ### \n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "\n",
      "Consensus: CTAGCTTGTACC\n",
      "Number of cliques: 6\n",
      "Best score of this group: 40\n",
      "Average score of this group: 41.5\n",
      "Motifs:\n",
      "Rank: 1, Score: 40\n",
      "Seq: 0, Position: 87, CGAGCTGGTACC\n",
      "Seq: 1, Position: 126, GTAGCTTGTACC\n",
      "Seq: 2, Position: 136, CTAGCCTGTACC\n",
      "Seq: 3, Position: 69, CTCGCTTGAGCC\n",
      "Seq: 4, Position: 183, CAAGCTTGTCCA\n",
      "Seq: 5, Position: 87, CTAGCTTGTACG\n",
      "Seq: 6, Position: 19, CTAGTTTGTACC\n",
      "Seq: 7, Position: 3, CTAGCTTTAACC\n",
      "Seq: 8, Position: 49, CTAGCTTGTATC\n",
      "Seq: 9, Position: 27, CTTGCTTTTACT\n",
      "Seq: 10, Position: 116, CTAGCTTCTACG\n",
      "Seq: 11, Position: 105, CTATCTTGTACT\n",
      "Seq: 12, Position: 134, CGCGATTGTACC\n",
      "Seq: 13, Position: 94, CTACCCGGTACC\n",
      "Seq: 14, Position: 38, CTAGCTCGTAGC\n",
      "Seq: 15, Position: 181, TTGGCGTGTACC\n",
      "Seq: 16, Position: 153, CTAGCTTGTATG\n",
      "Seq: 17, Position: 161, TTAGCCTGGACC\n",
      "Seq: 18, Position: 184, CTAGCATGTACC\n",
      "Seq: 19, Position: 187, CTAGATTGTACC\n",
      "\n",
      "Rank: 2, Score: 41\n",
      "Seq: 0, Position: 87, CGAGCTGGTACC\n",
      "Seq: 1, Position: 126, GTAGCTTGTACC\n",
      "Seq: 2, Position: 136, CTAGCCTGTACC\n",
      "Seq: 3, Position: 69, CTCGCTTGAGCC\n",
      "Seq: 4, Position: 150, CTAGGCTGAACT\n",
      "Seq: 5, Position: 87, CTAGCTTGTACG\n",
      "Seq: 6, Position: 19, CTAGTTTGTACC\n",
      "Seq: 7, Position: 3, CTAGCTTTAACC\n",
      "Seq: 8, Position: 49, CTAGCTTGTATC\n",
      "Seq: 9, Position: 27, CTTGCTTTTACT\n",
      "Seq: 10, Position: 116, CTAGCTTCTACG\n",
      "Seq: 11, Position: 105, CTATCTTGTACT\n",
      "Seq: 12, Position: 134, CGCGATTGTACC\n",
      "Seq: 13, Position: 94, CTACCCGGTACC\n",
      "Seq: 14, Position: 38, CTAGCTCGTAGC\n",
      "Seq: 15, Position: 181, TTGGCGTGTACC\n",
      "Seq: 16, Position: 153, CTAGCTTGTATG\n",
      "Seq: 17, Position: 161, TTAGCCTGGACC\n",
      "Seq: 18, Position: 184, CTAGCATGTACC\n",
      "Seq: 19, Position: 187, CTAGATTGTACC\n",
      "\n",
      "Rank: 2, Score: 41\n",
      "Seq: 0, Position: 87, CGAGCTGGTACC\n",
      "Seq: 1, Position: 126, GTAGCTTGTACC\n",
      "Seq: 2, Position: 136, CTAGCCTGTACC\n",
      "Seq: 3, Position: 69, CTCGCTTGAGCC\n",
      "Seq: 4, Position: 183, CAAGCTTGTCCA\n",
      "Seq: 5, Position: 35, CTACCTTGTACT\n",
      "Seq: 6, Position: 19, CTAGTTTGTACC\n",
      "Seq: 7, Position: 3, CTAGCTTTAACC\n",
      "Seq: 8, Position: 49, CTAGCTTGTATC\n",
      "Seq: 9, Position: 27, CTTGCTTTTACT\n",
      "Seq: 10, Position: 116, CTAGCTTCTACG\n",
      "Seq: 11, Position: 105, CTATCTTGTACT\n",
      "Seq: 12, Position: 134, CGCGATTGTACC\n",
      "Seq: 13, Position: 94, CTACCCGGTACC\n",
      "Seq: 14, Position: 38, CTAGCTCGTAGC\n",
      "Seq: 15, Position: 181, TTGGCGTGTACC\n",
      "Seq: 16, Position: 153, CTAGCTTGTATG\n",
      "Seq: 17, Position: 161, TTAGCCTGGACC\n",
      "Seq: 18, Position: 184, CTAGCATGTACC\n",
      "Seq: 19, Position: 187, CTAGATTGTACC\n",
      "\n",
      "Rank: 3, Score: 42\n",
      "Seq: 0, Position: 87, CGAGCTGGTACC\n",
      "Seq: 1, Position: 126, GTAGCTTGTACC\n",
      "Seq: 2, Position: 136, CTAGCCTGTACC\n",
      "Seq: 3, Position: 69, CTCGCTTGAGCC\n",
      "Seq: 4, Position: 150, CTAGGCTGAACT\n",
      "Seq: 5, Position: 35, CTACCTTGTACT\n",
      "Seq: 6, Position: 19, CTAGTTTGTACC\n",
      "Seq: 7, Position: 3, CTAGCTTTAACC\n",
      "Seq: 8, Position: 49, CTAGCTTGTATC\n",
      "Seq: 9, Position: 27, CTTGCTTTTACT\n",
      "Seq: 10, Position: 116, CTAGCTTCTACG\n",
      "Seq: 11, Position: 105, CTATCTTGTACT\n",
      "Seq: 12, Position: 134, CGCGATTGTACC\n",
      "Seq: 13, Position: 94, CTACCCGGTACC\n",
      "Seq: 14, Position: 38, CTAGCTCGTAGC\n",
      "Seq: 15, Position: 181, TTGGCGTGTACC\n",
      "Seq: 16, Position: 153, CTAGCTTGTATG\n",
      "Seq: 17, Position: 161, TTAGCCTGGACC\n",
      "Seq: 18, Position: 184, CTAGCATGTACC\n",
      "Seq: 19, Position: 187, CTAGATTGTACC\n",
      "\n",
      "Rank: 3, Score: 42\n",
      "Seq: 0, Position: 179, AACGCATGTACC\n",
      "Seq: 1, Position: 126, GTAGCTTGTACC\n",
      "Seq: 2, Position: 136, CTAGCCTGTACC\n",
      "Seq: 3, Position: 69, CTCGCTTGAGCC\n",
      "Seq: 4, Position: 183, CAAGCTTGTCCA\n",
      "Seq: 5, Position: 87, CTAGCTTGTACG\n",
      "Seq: 6, Position: 19, CTAGTTTGTACC\n",
      "Seq: 7, Position: 3, CTAGCTTTAACC\n",
      "Seq: 8, Position: 49, CTAGCTTGTATC\n",
      "Seq: 9, Position: 27, CTTGCTTTTACT\n",
      "Seq: 10, Position: 116, CTAGCTTCTACG\n",
      "Seq: 11, Position: 105, CTATCTTGTACT\n",
      "Seq: 12, Position: 134, CGCGATTGTACC\n",
      "Seq: 13, Position: 94, CTACCCGGTACC\n",
      "Seq: 14, Position: 38, CTAGCTCGTAGC\n",
      "Seq: 15, Position: 181, TTGGCGTGTACC\n",
      "Seq: 16, Position: 153, CTAGCTTGTATG\n",
      "Seq: 17, Position: 161, TTAGCCTGGACC\n",
      "Seq: 18, Position: 184, CTAGCATGTACC\n",
      "Seq: 19, Position: 187, CTAGATTGTACC\n",
      "\n",
      "Rank: 4, Score: 43\n",
      "Seq: 0, Position: 179, AACGCATGTACC\n",
      "Seq: 1, Position: 126, GTAGCTTGTACC\n",
      "Seq: 2, Position: 136, CTAGCCTGTACC\n",
      "Seq: 3, Position: 69, CTCGCTTGAGCC\n",
      "Seq: 4, Position: 183, CAAGCTTGTCCA\n",
      "Seq: 5, Position: 35, CTACCTTGTACT\n",
      "Seq: 6, Position: 19, CTAGTTTGTACC\n",
      "Seq: 7, Position: 3, CTAGCTTTAACC\n",
      "Seq: 8, Position: 49, CTAGCTTGTATC\n",
      "Seq: 9, Position: 27, CTTGCTTTTACT\n",
      "Seq: 10, Position: 116, CTAGCTTCTACG\n",
      "Seq: 11, Position: 105, CTATCTTGTACT\n",
      "Seq: 12, Position: 134, CGCGATTGTACC\n",
      "Seq: 13, Position: 94, CTACCCGGTACC\n",
      "Seq: 14, Position: 38, CTAGCTCGTAGC\n",
      "Seq: 15, Position: 181, TTGGCGTGTACC\n",
      "Seq: 16, Position: 153, CTAGCTTGTATG\n",
      "Seq: 17, Position: 161, TTAGCCTGGACC\n",
      "Seq: 18, Position: 184, CTAGCATGTACC\n",
      "Seq: 19, Position: 187, CTAGATTGTACC\n",
      "\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "\n",
      "Total Cliques: 6\n",
      "Total Consensus Motifs: 1\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "\n",
      "Top 3 Motifs\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "\n",
      "Consensus: CTAGCTTGTACC, Rank: 1, Score: 40\n",
      "Instances: \n",
      "Seq: 0, Position: 87, CGAGCTGGTACC\n",
      "Seq: 1, Position: 126, GTAGCTTGTACC\n",
      "Seq: 2, Position: 136, CTAGCCTGTACC\n",
      "Seq: 3, Position: 69, CTCGCTTGAGCC\n",
      "Seq: 4, Position: 183, CAAGCTTGTCCA\n",
      "Seq: 5, Position: 87, CTAGCTTGTACG\n",
      "Seq: 6, Position: 19, CTAGTTTGTACC\n",
      "Seq: 7, Position: 3, CTAGCTTTAACC\n",
      "Seq: 8, Position: 49, CTAGCTTGTATC\n",
      "Seq: 9, Position: 27, CTTGCTTTTACT\n",
      "Seq: 10, Position: 116, CTAGCTTCTACG\n",
      "Seq: 11, Position: 105, CTATCTTGTACT\n",
      "Seq: 12, Position: 134, CGCGATTGTACC\n",
      "Seq: 13, Position: 94, CTACCCGGTACC\n",
      "Seq: 14, Position: 38, CTAGCTCGTAGC\n",
      "Seq: 15, Position: 181, TTGGCGTGTACC\n",
      "Seq: 16, Position: 153, CTAGCTTGTATG\n",
      "Seq: 17, Position: 161, TTAGCCTGGACC\n",
      "Seq: 18, Position: 184, CTAGCATGTACC\n",
      "Seq: 19, Position: 187, CTAGATTGTACC\n",
      "\n",
      "Consensus: CTAGCTTGTACC, Rank: 2, Score: 41\n",
      "Instances: \n",
      "Seq: 0, Position: 87, CGAGCTGGTACC\n",
      "Seq: 1, Position: 126, GTAGCTTGTACC\n",
      "Seq: 2, Position: 136, CTAGCCTGTACC\n",
      "Seq: 3, Position: 69, CTCGCTTGAGCC\n",
      "Seq: 4, Position: 150, CTAGGCTGAACT\n",
      "Seq: 5, Position: 87, CTAGCTTGTACG\n",
      "Seq: 6, Position: 19, CTAGTTTGTACC\n",
      "Seq: 7, Position: 3, CTAGCTTTAACC\n",
      "Seq: 8, Position: 49, CTAGCTTGTATC\n",
      "Seq: 9, Position: 27, CTTGCTTTTACT\n",
      "Seq: 10, Position: 116, CTAGCTTCTACG\n",
      "Seq: 11, Position: 105, CTATCTTGTACT\n",
      "Seq: 12, Position: 134, CGCGATTGTACC\n",
      "Seq: 13, Position: 94, CTACCCGGTACC\n",
      "Seq: 14, Position: 38, CTAGCTCGTAGC\n",
      "Seq: 15, Position: 181, TTGGCGTGTACC\n",
      "Seq: 16, Position: 153, CTAGCTTGTATG\n",
      "Seq: 17, Position: 161, TTAGCCTGGACC\n",
      "Seq: 18, Position: 184, CTAGCATGTACC\n",
      "Seq: 19, Position: 187, CTAGATTGTACC\n",
      "\n",
      "Consensus: CTAGCTTGTACC, Rank: 2, Score: 41\n",
      "Instances: \n",
      "Seq: 0, Position: 87, CGAGCTGGTACC\n",
      "Seq: 1, Position: 126, GTAGCTTGTACC\n",
      "Seq: 2, Position: 136, CTAGCCTGTACC\n",
      "Seq: 3, Position: 69, CTCGCTTGAGCC\n",
      "Seq: 4, Position: 183, CAAGCTTGTCCA\n",
      "Seq: 5, Position: 35, CTACCTTGTACT\n",
      "Seq: 6, Position: 19, CTAGTTTGTACC\n",
      "Seq: 7, Position: 3, CTAGCTTTAACC\n",
      "Seq: 8, Position: 49, CTAGCTTGTATC\n",
      "Seq: 9, Position: 27, CTTGCTTTTACT\n",
      "Seq: 10, Position: 116, CTAGCTTCTACG\n",
      "Seq: 11, Position: 105, CTATCTTGTACT\n",
      "Seq: 12, Position: 134, CGCGATTGTACC\n",
      "Seq: 13, Position: 94, CTACCCGGTACC\n",
      "Seq: 14, Position: 38, CTAGCTCGTAGC\n",
      "Seq: 15, Position: 181, TTGGCGTGTACC\n",
      "Seq: 16, Position: 153, CTAGCTTGTATG\n",
      "Seq: 17, Position: 161, TTAGCCTGGACC\n",
      "Seq: 18, Position: 184, CTAGCATGTACC\n",
      "Seq: 19, Position: 187, CTAGATTGTACC\n",
      "\n"
     ]
    }
   ],
   "execution_count": 26
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-04-20T20:47:56.947039Z",
     "start_time": "2024-04-20T20:47:56.943591Z"
    }
   },
   "cell_type": "code",
   "source": [
    "for best_motif in _12_3_found_motif:\n",
    "    try:\n",
    "        assert best_motif == implanted_motif\n",
    "        print(\"Success\")\n",
    "    except AssertionError:\n",
    "        print(\"Fail\")\n",
    "        continue\n",
    "print(implanted_motif)"
   ],
   "id": "b012499eca45db26",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Success\n",
      "Success\n",
      "Success\n",
      "CTAGCTTGTACC\n"
     ]
    }
   ],
   "execution_count": 27
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "### Test with larger sequence length",
   "id": "7e865d83b09bb64a"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "dataset, implanted_motif, instances = generate_15_4_dataset(600, 15, 4)\n",
    "large_set_found_motif = find_motif(dataset, 15, 4, output_consensus_groups=True)\n"
   ],
   "id": "27b0b5ccfd6f467e",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "\n",
    "for best_motif in large_set_found_motif:\n",
    "    try:\n",
    "        assert best_motif == implanted_motif\n",
    "        print(\"Success\")\n",
    "    except AssertionError:\n",
    "        print(\"Fail\")\n",
    "        continue\n",
    "print(implanted_motif)"
   ],
   "id": "49102568ef00784a",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "### Test with 50 15,4 datasets ###\n",
    "\n",
    "To replicate Sun et al's testing scheme as closely as possible, test the algorithm with 50 different datsets.\n",
    "\n",
    "First try a fast one with a sequence length of 100\n",
    "\n",
    "\n"
   ],
   "id": "e5a763fccf4220d3"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-04-21T02:07:17.602401Z",
     "start_time": "2024-04-21T02:06:50.694714Z"
    }
   },
   "cell_type": "code",
   "source": [
    "implanted_motif_list = []\n",
    "found_motif_list = []\n",
    "for i in tqdm(range(50), desc='Testing datasets', position=0, leave=True):\n",
    "    dataset, implanted_motif, instances = generate_15_4_dataset(80, 15, 4)\n",
    "    found_motif = find_motif(dataset, 15, 4, print_output=False, show_progress=False)\n",
    "    \n",
    "    implanted_motif_list.append(implanted_motif)\n",
    "    found_motif_list.append(found_motif)\n",
    "    "
   ],
   "id": "35cffb2a488c1f6b",
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Testing datasets: 100%|██████████| 50/50 [00:26<00:00,  1.86it/s]\n"
     ]
    }
   ],
   "execution_count": 26
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-04-21T02:08:53.602933Z",
     "start_time": "2024-04-21T02:08:53.597869Z"
    }
   },
   "cell_type": "code",
   "source": [
    "print(\"15,4 Implanted Motif Tests\\n\")\n",
    "for i in range(50):\n",
    "    print(\"Test \" + str(i + 1))\n",
    "    print(\"Implanted Motif: \" + implanted_motif_list[i])\n",
    "    for j in range(len(found_motif_list[i])):\n",
    "        match = \"\"\n",
    "        if implanted_motif_list[i] == found_motif_list[i][j]:\n",
    "            match = \"\\tMATCH\"\n",
    "        print(\"Predicted Motif \" + str(j + 1) + \": \" + found_motif_list[i][j] + match)\n",
    "    print()"
   ],
   "id": "2229c030833d41c9",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15,4 Implanted Motif Tests\n",
      "\n",
      "Test 1\n",
      "Implanted Motif: GTCCGGAACCCCCGT\n",
      "Predicted Motif 1: GTCCGGAACCCCCGT\tMATCH\n",
      "Predicted Motif 2: GTCCGGAACCCCCGT\tMATCH\n",
      "\n",
      "Test 2\n",
      "Implanted Motif: CGCACTACAGACACG\n",
      "Predicted Motif 1: CGCACTACAGACACG\tMATCH\n",
      "Predicted Motif 2: GCACTACAGACACGG\n",
      "\n",
      "Test 3\n",
      "Implanted Motif: AAAAGCAATCGTTGA\n",
      "Predicted Motif 1: AAAAGCAATCGTTGA\tMATCH\n",
      "Predicted Motif 2: TAAAAGCAATCGTTG\n",
      "\n",
      "Test 4\n",
      "Implanted Motif: TAGCTACCTAAATCT\n",
      "Predicted Motif 1: TAGCTACCTAAATCT\tMATCH\n",
      "Predicted Motif 2: TTAGCTACCTAAATC\n",
      "Predicted Motif 3: TTAGCTACCTAAATC\n",
      "\n",
      "Test 5\n",
      "Implanted Motif: CTGCGTACATTCATA\n",
      "Predicted Motif 1: CTGCGTACATTCATA\tMATCH\n",
      "Predicted Motif 2: GCTGCGTACATTCAT\n",
      "Predicted Motif 3: GCTGCGTACATTCAT\n",
      "\n",
      "Test 6\n",
      "Implanted Motif: TACCCCGGGCCGTCC\n",
      "Predicted Motif 1: TACCCCGGGCCGTCC\tMATCH\n",
      "\n",
      "Test 7\n",
      "Implanted Motif: AGCCTACTTATTTTG\n",
      "Predicted Motif 1: AGCCTACTTATTTTG\tMATCH\n",
      "Predicted Motif 2: GCCTACTTATTTTGA\n",
      "\n",
      "Test 8\n",
      "Implanted Motif: CCTTCCAAGAATGGA\n",
      "Predicted Motif 1: CCTTCCAAGAATGGA\tMATCH\n",
      "Predicted Motif 2: CTTCCAAGAATGGAA\n",
      "\n",
      "Test 9\n",
      "Implanted Motif: CATTGTATCCCATGA\n",
      "Predicted Motif 1: CATTGTATCCCATGA\tMATCH\n",
      "Predicted Motif 2: ATTGTATCCCATGAC\n",
      "\n",
      "Test 10\n",
      "Implanted Motif: CGCGGAGCGACCGCC\n",
      "Predicted Motif 1: CGCGGAGCGACCGCC\tMATCH\n",
      "Predicted Motif 2: GCGGAGCGACCGCCA\n",
      "Predicted Motif 3: GCGGAGCGACCGCCA\n",
      "\n",
      "Test 11\n",
      "Implanted Motif: CCGACTCCTTCCGTA\n",
      "Predicted Motif 1: CCGACTCCTTCCGTA\tMATCH\n",
      "Predicted Motif 2: CCGACTCCTTCCGTA\tMATCH\n",
      "Predicted Motif 3: CCGACTCCTTCCGTA\tMATCH\n",
      "\n",
      "Test 12\n",
      "Implanted Motif: CATGATCCCCAACTG\n",
      "Predicted Motif 1: CATGATCCCCAACTG\tMATCH\n",
      "\n",
      "Test 13\n",
      "Implanted Motif: GCGGGAGACACAGAC\n",
      "Predicted Motif 1: GCGGGAGACACAGAC\tMATCH\n",
      "Predicted Motif 2: GCGGGAGACACAGAC\tMATCH\n",
      "Predicted Motif 3: TGCGGGAGACACAGA\n",
      "\n",
      "Test 14\n",
      "Implanted Motif: GGTCATTCCTCGCGC\n",
      "Predicted Motif 1: GGTCATTCCTCGCGC\tMATCH\n",
      "Predicted Motif 2: AGGTCATTCCTCGCG\n",
      "Predicted Motif 3: GTCATTCCTCGCGCA\n",
      "\n",
      "Test 15\n",
      "Implanted Motif: CTGCGAGAGATCAGA\n",
      "Predicted Motif 1: CTGCGAGAGATCAGA\tMATCH\n",
      "\n",
      "Test 16\n",
      "Implanted Motif: GCTGAGATTTAGGAC\n",
      "Predicted Motif 1: GCTGAGATTTAGGAC\tMATCH\n",
      "Predicted Motif 2: GCTGAGATTTAGGAC\tMATCH\n",
      "\n",
      "Test 17\n",
      "Implanted Motif: TAAAGGGGCCATCGT\n",
      "Predicted Motif 1: AAAGGGGCCATCGTG\n",
      "\n",
      "Test 18\n",
      "Implanted Motif: GGGTGGGTTAGCCAA\n",
      "Predicted Motif 1: GGGTGGGTTAGCCAA\tMATCH\n",
      "Predicted Motif 2: GGGTGGGTTAGCCAA\tMATCH\n",
      "Predicted Motif 3: CGGGTGGGTTAGCCA\n",
      "\n",
      "Test 19\n",
      "Implanted Motif: TCAACATCGCCATAT\n",
      "Predicted Motif 1: TCAACATCGCCATAT\tMATCH\n",
      "Predicted Motif 2: CTCAACATCGCCATA\n",
      "\n",
      "Test 20\n",
      "Implanted Motif: CCGGGGCAGGTTCCT\n",
      "Predicted Motif 1: CCGGGGCAGGTTCCT\tMATCH\n",
      "\n",
      "Test 21\n",
      "Implanted Motif: ATGGGCTGCGATCTC\n",
      "Predicted Motif 1: ATGGGCTGCGATCTC\tMATCH\n",
      "\n",
      "Test 22\n",
      "Implanted Motif: AATATTTTTGGTATG\n",
      "Predicted Motif 1: AATATTTTTGGTATG\tMATCH\n",
      "Predicted Motif 2: AATATTTTTGGTATG\tMATCH\n",
      "\n",
      "Test 23\n",
      "Implanted Motif: ACTTTTGCATTCTTC\n",
      "Predicted Motif 1: ACTTTTGCATTCTTC\tMATCH\n",
      "Predicted Motif 2: ACTTTTGCATTCTTC\tMATCH\n",
      "Predicted Motif 3: CTTTTGCATTCTTCT\n",
      "\n",
      "Test 24\n",
      "Implanted Motif: GATTGGGATGGTACA\n",
      "Predicted Motif 1: GATTGGGATGGTACA\tMATCH\n",
      "Predicted Motif 2: ATTGGGATGGTACAT\n",
      "Predicted Motif 3: GATTGGGATGGTACA\tMATCH\n",
      "\n",
      "Test 25\n",
      "Implanted Motif: GGACATGGGGTGCAC\n",
      "Predicted Motif 1: GGACATGGGGTGCAC\tMATCH\n",
      "Predicted Motif 2: AGGACATGGGGTGCA\n",
      "\n",
      "Test 26\n",
      "Implanted Motif: GGAGCGACGATTAAA\n",
      "Predicted Motif 1: GGAGCGACGATTAAA\tMATCH\n",
      "Predicted Motif 2: TGGAGCGACGATTAA\n",
      "\n",
      "Test 27\n",
      "Implanted Motif: AACCTGCACAGTTTG\n",
      "Predicted Motif 1: AACCTGCACAGTTTG\tMATCH\n",
      "Predicted Motif 2: AACCTGCACAGTTTG\tMATCH\n",
      "\n",
      "Test 28\n",
      "Implanted Motif: TTAGGCCTTATAGCG\n",
      "Predicted Motif 1: TTAGGCCTTATAGCG\tMATCH\n",
      "Predicted Motif 2: TTAGGCCTTATAGCG\tMATCH\n",
      "Predicted Motif 3: TAGGCCTTATAGCGG\n",
      "\n",
      "Test 29\n",
      "Implanted Motif: CCTTTGAACGCATAT\n",
      "Predicted Motif 1: CCTTTGAACGCATAT\tMATCH\n",
      "Predicted Motif 2: CCTTTGAACGCATAT\tMATCH\n",
      "Predicted Motif 3: CTTTGAACGCATATT\n",
      "\n",
      "Test 30\n",
      "Implanted Motif: TTCGGCCAATGTGTC\n",
      "Predicted Motif 1: TTCGGCCAATGTGTC\tMATCH\n",
      "Predicted Motif 2: TTCGGCCAATGTGTC\tMATCH\n",
      "\n",
      "Test 31\n",
      "Implanted Motif: CCAGACTAGTCGATC\n",
      "Predicted Motif 1: CCAGACTAGTCGATC\tMATCH\n",
      "Predicted Motif 2: CCAGACTAGTCGATC\tMATCH\n",
      "Predicted Motif 3: CCAGACTAGTCGATC\tMATCH\n",
      "\n",
      "Test 32\n",
      "Implanted Motif: CCTCTTGATGAGGCT\n",
      "Predicted Motif 1: CCTCTTGATGAGGCT\tMATCH\n",
      "Predicted Motif 2: CCTCTTGATGAGGCT\tMATCH\n",
      "Predicted Motif 3: CCTCTTGATGAGGCT\tMATCH\n",
      "\n",
      "Test 33\n",
      "Implanted Motif: CGCAACAGGAGTTTA\n",
      "Predicted Motif 1: CGCAACAGGAGTTTA\tMATCH\n",
      "Predicted Motif 2: TCGCAACAGGAGTTT\n",
      "\n",
      "Test 34\n",
      "Implanted Motif: TGCGGGGCTTAAATT\n",
      "Predicted Motif 1: TGCGGGGCTTAAATT\tMATCH\n",
      "Predicted Motif 2: TGCGGGGCTTAAATT\tMATCH\n",
      "Predicted Motif 3: GCGGGGCTTAAATTT\n",
      "\n",
      "Test 35\n",
      "Implanted Motif: CGAAGCGAATGCAGC\n",
      "Predicted Motif 1: CGAAGCGAATGCAGC\tMATCH\n",
      "Predicted Motif 2: GAAGCGAATGCAGCT\n",
      "\n",
      "Test 36\n",
      "Implanted Motif: CGGTAGCACTGATTG\n",
      "Predicted Motif 1: CGGTAGCACTGATTG\tMATCH\n",
      "Predicted Motif 2: ACGGTAGCACTGATT\n",
      "Predicted Motif 3: GGTAGCACTGATTGG\n",
      "\n",
      "Test 37\n",
      "Implanted Motif: AGCTGACATTGTGGG\n",
      "Predicted Motif 1: AGCTGACATTGTGGG\tMATCH\n",
      "Predicted Motif 2: GCTGACATTGTGGGG\n",
      "\n",
      "Test 38\n",
      "Implanted Motif: GTAAGACTTCGTTTC\n",
      "Predicted Motif 1: GTAAGACTTCGTTTC\tMATCH\n",
      "Predicted Motif 2: GTAAGACTTCGTTTC\tMATCH\n",
      "Predicted Motif 3: GGTAAGACTTCGTTT\n",
      "\n",
      "Test 39\n",
      "Implanted Motif: AGAGAGCCCTTAGCC\n",
      "Predicted Motif 1: AGAGAGCCCTTAGCC\tMATCH\n",
      "Predicted Motif 2: AGAGAGCCCTTAGCC\tMATCH\n",
      "Predicted Motif 3: AGAGAGCCCTTAGCC\tMATCH\n",
      "\n",
      "Test 40\n",
      "Implanted Motif: AGTCACCACGGGTTC\n",
      "Predicted Motif 1: AGTCACCACGGGTTC\tMATCH\n",
      "Predicted Motif 2: GTCACCACGGGTTCT\n",
      "Predicted Motif 3: TCACCACGGGTTCTG\n",
      "\n",
      "Test 41\n",
      "Implanted Motif: TTCTGCGGTTGCTAC\n",
      "Predicted Motif 1: TTCTGCGGTTGCTAC\tMATCH\n",
      "Predicted Motif 2: TTCTGCGGTTGCTAC\tMATCH\n",
      "Predicted Motif 3: TTCTGCGGTTGCTAC\tMATCH\n",
      "\n",
      "Test 42\n",
      "Implanted Motif: TCATCCCCGAGCTAA\n",
      "Predicted Motif 1: TCATCCCCGAGCTAA\tMATCH\n",
      "Predicted Motif 2: TCATCCCCGAGCTAA\tMATCH\n",
      "Predicted Motif 3: TCATCCCCGAGCTAA\tMATCH\n",
      "\n",
      "Test 43\n",
      "Implanted Motif: TCATCATTCCGACTT\n",
      "Predicted Motif 1: TCATCATTCCGACTT\tMATCH\n",
      "Predicted Motif 2: CATCATTCCGACTTT\n",
      "Predicted Motif 3: GTCATCATTCCGACT\n",
      "\n",
      "Test 44\n",
      "Implanted Motif: CAGCCAACTGAGAAT\n",
      "Predicted Motif 1: CAGCCAACTGAGAAT\tMATCH\n",
      "Predicted Motif 2: CAGCCAACTGAGAAT\tMATCH\n",
      "Predicted Motif 3: GCAGCCAACTGAGAA\n",
      "\n",
      "Test 45\n",
      "Implanted Motif: AGTCAACATGCGCAG\n",
      "Predicted Motif 1: AGTCAACATGCGCAG\tMATCH\n",
      "Predicted Motif 2: AGTCAACATGCGCAG\tMATCH\n",
      "\n",
      "Test 46\n",
      "Implanted Motif: CTCATGCCCGCGTTG\n",
      "Predicted Motif 1: CTCATGCCCGCGTTG\tMATCH\n",
      "\n",
      "Test 47\n",
      "Implanted Motif: CTAGGGACCTACGCT\n",
      "Predicted Motif 1: CTAGGGACCTACGCT\tMATCH\n",
      "Predicted Motif 2: TCTAGGGACCTACGC\n",
      "Predicted Motif 3: ATCTAGGGACCTACG\n",
      "\n",
      "Test 48\n",
      "Implanted Motif: CACTAGACTGCTACT\n",
      "Predicted Motif 1: CACTAGACTGCTACT\tMATCH\n",
      "Predicted Motif 2: ACTAGACTGCTACTA\n",
      "Predicted Motif 3: GCACTAGACTGCTAC\n",
      "\n",
      "Test 49\n",
      "Implanted Motif: ATGACCGCTTCGCTA\n",
      "Predicted Motif 1: ATGACCGCTTCGCTA\tMATCH\n",
      "\n",
      "Test 50\n",
      "Implanted Motif: TGAAGTTTTGCCATA\n",
      "Predicted Motif 1: TGAAGTTTTGCCATA\tMATCH\n",
      "Predicted Motif 2: TGAAGTTTTGCCATA\tMATCH\n",
      "Predicted Motif 3: TGAAGTTTTGCCATA\tMATCH\n",
      "\n"
     ]
    }
   ],
   "execution_count": 28
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "# Evaluation\n",
    "\n",
    "### Replicate the evaluation scheme of Sun et al\n",
    "\n",
    "Test 50 datasets of 20 600-bp long sequences for each of four (l,d) motifs:\n",
    "\n",
    "* 12,3\n",
    "* 13,3\n",
    "* 14,4\n",
    "* 15,4\n",
    "\n",
    "note: Sun et al's evaluation included additional l,d motifs that we do not include in our tests.  \n",
    "\n",
    "The results are evaluated with nucleotide based metrics introduced by Pevzner and Sze (2000)  \n",
    "\n",
    "The metrics, nucleotide performance coefficient (nPC) and nucleotide sensitivity (nSn) are built from a nucleotide level confusion matrix, where:  \n",
    "\n",
    "* True Positives, TP is the number of nucleotide positions in both known sites and predicted sites\n",
    "* False Negatives, FN is the number of nucleotide positions in known sites but not in predicted sites\n",
    "* False Positive, FP is the number of nucleotide positions not in known sites but in predicted sites \n",
    "* True Negative, TN is the number of nucleotide positions in neither known sites nor predicted sites\n",
    "\n",
    "\n",
    "&emsp;&emsp;nPC = TP/(TP+FN+FP)  \n",
    "&emsp;&emsp;nSn = TP/(TP+FN)  <p></p>\n",
    "\n",
    "The best motif of the top 3 ranked motifs was used to obtain the best metric. Note that 0, 1, or 2 motifs is also a possible result.\n",
    "\n",
    "These metrics were averaged over all 50 tests. Sun et al only used nSn in a comparison with other algorithms, so the only result available is for 15,4 motifs.\n",
    "\n",
    "The following results were presented by Sun et al: <p></p>\n",
    "\n",
    "| l,d  | nPC         | nSn         |\n",
    "|------|-------------|-------------|\n",
    "| 12,3 | 0.82 ± 0.07 | -           |\n",
    "| 13,3 | 0.94 ± 0.05 | -           |\n",
    "| 14,4 | 0.80 ± 0.19 | -           |\n",
    "| 15,4 | 0.95 ± 0.05 | 1.00 ± 0.01 |\n",
    "\n",
    "Due to time constraints for submitting the project we did not run this evaluation in the notebook. \n",
    "\n",
    "Separate python scripts were generated to run each test in unison on NCSU's Hazel cluster, and metrics were calculated by hand based on the output shown in the cell above. <p></p>\n",
    "\n",
    "For example, if one of the 50 tests had this result: \n",
    "\n",
    "    Test 3  \n",
    "    Implanted Motif: GAATCAGGGTGAA  \n",
    "    Predicted Motif 1: GAATCAGGGTGAA    MATCH  \n",
    "    Predicted Motif 2: GAATCAGGGTGAA    MATCH  \n",
    "    Predicted Motif 3: GAATCAGGGTGAA    MATCH  \n",
    "\n",
    "or this result\n",
    "\n",
    "    Test 7  \n",
    "    Implanted Motif: GTTGGCAGTCGGA  \n",
    "    Predicted Motif 1: GTTGGCAGTCGGA    MATCH  \n",
    "    Predicted Motif 2: TTGGCAGTCGGAG  \n",
    "\n",
    "then nPC = 1. If the motif was correctly identified, there are no FNs or FPs. <p></p>\n",
    "\n",
    "However, for this result:\n",
    "\n",
    "    Test 50  \n",
    "    Implanted Motif: TCCATAGGAGATT  \n",
    "    Predicted Motif 1: CCATAGGAGATTT  \n",
    "    Predicted Motif 2: CCATAGGAGATTT  \n",
    "    Predicted Motif 3: CCATAGGAGATTT  \n",
    "\n",
    "the motif was not precisely identified, resulting in nucleotide FNs and FPs.\n",
    "  \n",
    "In this case, the implanted motif and predicted motif are misaligned by 1 base.\n",
    "\n",
    "There is a FN before the beginning of the predicted motif, and a FP at the last base. \n",
    "\n",
    " TCCATAGGAGATT  \n",
    " <span style=\"color: red;\">T</span>CCATAGGAGATT<span style=\"color: orange;\">T</span>\n",
    "\n",
    "Therefore, nPC = 13/(13+1+1) = 0.87 <p></p>\n",
    "\n",
    "Finally, for this result:\n",
    "\n",
    "    Test 19  \n",
    "    Implanted Motif: ATTGTAATTCACT  \n",
    "    Predicted Motif 1: no motifs found :(  \n",
    "\n",
    "nPC = 0.\n",
    "\n"
   ],
   "id": "3635a021063d3962"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "# Test with the Sandve datasets",
   "id": "2fdb705c7979a0cb"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "def read_sandve_fasta(filepath):\n",
    "    sequences = {}\n",
    "    current_name = None\n",
    "    seq_id = 0\n",
    "    \n",
    "    with open(filepath, 'r') as file:\n",
    "        for line in file:\n",
    "            line = line.strip()\n",
    "            if line.startswith('>'):\n",
    "                current_name = line[1:]\n",
    "                match = re.search(r'-(\\d+)-', current_name)\n",
    "                seq_id = int(match.group(1))\n",
    "            else:\n",
    "                sequences[seq_id] = line\n",
    "                \n",
    "    \n",
    "    sorted_len = dict(sorted(sequences.items(), key=lambda x: len(x[1]), reverse=True))\n",
    "    order = list(sorted_len.keys())\n",
    "    sorted_seqs = list(sorted_len.values())\n",
    "    \n",
    "    return order, sorted_seqs\n",
    "    \n",
    "    "
   ],
   "id": "90172421c74a1946"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "folder = \"sandve_datasets/algorithm_real/\"\n",
    "fas_files = os.listdir(folder)\n",
    "\n",
    "print(fas_files[0])\n",
    "# order, sorted_seqs = read_sandve_fasta(folder + fas_files[0])\n",
    "order, sorted_seqs = read_sandve_fasta(folder + \"M00441.fas\")\n",
    "\n",
    "print(order)\n",
    "print(len(sorted_seqs))\n",
    "print(sorted_seqs)\n"
   ],
   "id": "95727ecf5ff2a3ac"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "sandve_result1 = find_motif(sorted_seqs, 8)",
   "id": "8cb104c2858ed2a7"
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
